core/src/main/java/org/apache/brooklyn/core/config/BasicConfigInheritance.java

```diff
@@ -28,51 +32,188 @@
+            return getDelegate().considerParent(local, parent, context);
+        }
+        @Override
+        public <TContainer, TValue> ReferenceWithError<ConfigValueAtContainer<TContainer, TValue>> resolveWithParent(ConfigValueAtContainer<TContainer, TValue> local, ConfigValueAtContainer<TContainer, TValue> resolvedParent, ConfigInheritanceContext context) {
+            return getDelegate().resolveWithParent(local, resolvedParent, context);
+        }
+        @Override
+        public boolean equals(Object obj) {
+            return super.equals(obj) || getDelegate().equals(obj);
+        }
+        
+        // standard deserialization method
+        protected ConfigInheritance readResolve() {
+            return returnEquivalentConstant(this);
+        }
+    }
+
+    private static ConfigInheritance returnEquivalentConstant(ConfigInheritance candidate) {
+        for (ConfigInheritance knownMode: Arrays.asList(
+                NOT_REINHERITED, NOT_REINHERITED_ELSE_DEEP_MERGE, NEVER_INHERITED, OVERWRITE, BasicConfigInheritance.DEEP_MERGE)) {
+            if (candidate.equals(knownMode)) return knownMode;
+        }
+        if (candidate.equals(new BasicConfigInheritance(false, CONFLICT_RESOLUTION_STRATEGY_OVERWRITE, true, true))) {
+            // ignore the ancestor flag for this mode
+            return NEVER_INHERITED;
+        }
+        return candidate;
+    }
+    
+    /*
+     * use of delegate is so that stateless classes can be defined to make the serialization nice,
+     * both the name and hiding the implementation detail (also making it easier for that detail to change);
+     * with aliased type names the field names here could even be aliases for the type names.
+     * (we could alternatively have an "alias-for-final-instance" mode in serialization,
+     * to optimize where we know that a java instance is final.)   
+     */
+    
+    private static class NotReinherited extends DelegatingConfigInheritance {
+        final static BasicConfigInheritance DELEGATE = new BasicConfigInheritance(false, CONFLICT_RESOLUTION_STRATEGY_OVERWRITE, false, true); 
+        @Override protected ConfigInheritance getDelegate() { return DELEGATE; }
+    }
     /** Indicates that a config key value should not be passed down from a container where it is defined.
      * Unlike {@link #NEVER_INHERITED} these values can be passed down if set as anonymous keys at a container
      * (ie the container does not expect it) to a container which does expect it, but it will not be passed down further. 
      * If the inheritor also defines a value the parent's value is ignored irrespective 
      * (as in {@link #OVERWRITE}; see {@link #NOT_REINHERITED_ELSE_DEEP_MERGE} if merging is desired). */
-    public static BasicConfigInheritance NOT_REINHERITED = new BasicConfigInheritance(false,CONFLICT_RESOLUTION_STRATEGY_OVERWRITE,false);
+    public static ConfigInheritance NOT_REINHERITED = new NotReinherited();
```

Comment:
@neykov my idea was for the class to such that if serialized it just writes the class name; won't your variant write all the fields passed to `BCI` ?